src/python/pants/core/util_rules/environments.py

```diff
@@ -0,0 +1,180 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+from pants.build_graph.address import Address, AddressInput
+from pants.engine.platform import Platform
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    StringSequenceField,
+    Target,
+    WrappedTarget,
+    WrappedTargetRequest,
+)
+from pants.option.option_types import DictOption
+from pants.option.subsystem import Subsystem
+from pants.util.memo import memoized_property
+from pants.util.strutil import softwrap
+
+
+class EnvironmentsSubsystem(Subsystem):
+    options_scope = "environments-preview"
+    help = softwrap(
+        """
+        A highly experimental subsystem to allow setting environment variables and executable
+        search paths for different environments, e.g. macOS vs. Linux.
+        """
+    )
+
+    _platforms_to_local_environment = DictOption[str](
+        help=softwrap(
+            """
+            A mapping of platform strings to addresses to `_local_environment` targets. For example:
+
+                [environments-preview.platforms_to_local_environment]
+                linux_arm64 = "//:linux_arm64_environment"
+                linux_x86_64 = "//:linux_x86_environment"
+                macos_arm64 = "//:macos_environment"
+                macos_x86_64 = "//:macos_environment"
+
+            Pants will detect what platform you are currently on and load the specified
+            environment. If a platform is not defined, then Pants will use legacy options like
+            `[python-bootstrap].search_path`.
+
+            Warning: this feature is experimental and this option may be changed and removed before
+            the first 2.15 alpha release.
+            """
+        )
+    )
+
+    @memoized_property
+    def platforms_to_local_environment(self) -> dict[str, str]:
+        valid_platforms = {plat.value for plat in Platform}
+        invalid_keys = set(self._platforms_to_local_environment.keys()) - valid_platforms
+        if invalid_keys:
+            raise ValueError(
+                softwrap(
+                    f"""
+                    Unrecognized platforms as the keys to the option
+                    `[environments-preview].platforms_to_local_environment`: {sorted(invalid_keys)}
+
+                    All valid platforms: {sorted(valid_platforms)}
+                    """
+                )
+            )
+        return self._platforms_to_local_environment
+
+
+# -------------------------------------------------------------------------------------------
+# Environment targets
+# -------------------------------------------------------------------------------------------
+
+
+class PythonInterpreterSearchPathsField(StringSequenceField):
+    alias = "python_interpreter_search_paths"
```

Comment:
We want to make sure we like this field name...Note that we used to have `[python-setup].interpreter_search_path`, then it became `[python-bootstrap].search_paths`. It looks like we use `paths` rather than `path` everywhere now.